sips/sip-021/sip-021-trustless-two-way-peg-for-bitcoin.md

```diff
@@ -0,0 +1,1385 @@
+is taken, miners are required to process it in any subsequent Stacks block on a
+fork that does not yet include it (thereby materializing the revelation in all
+forks).
+
+## Fork Rules
+
+This proposal preserves Stacks' Nakamoto consensus in order to remain resilient
+in the face of node churn and network unavailability.  The ability to fork
+gives the blockchain a degree of resilience not found in traditional replicated
+state machine systems (including PoS blockchains): there is no barrier-to-entry
+to become a block producer, and the system will converge on the same canonical
+fork history as long as there exists a BFT-majority of connected block
+producers.
+
+However, a key challenge to designing a 2-way peg between Bitcoin and a
+blockchain with Nakamoto consensus is that the blockchain can fork, potentially
+leading to inconsistencies between blockchains forks and Bitcoin forks.  Stacks
+addresses this challenge in two ways:
+
+* Stacks blocks are tied to Bitcoin blocks, so if Bitcoin forks, the Stacks
+blockchain's nodes will automatically track both forks.  This ensures that
+Bitcoin forks do not produce inconsistent Stacks forks.  This is the behavior
+of the system today.
+
+* Stacks maintains Nakamoto consensus for a fixed number
+of Bitcoin blocks.  That is, Stacks can fork independently of Bitcoin, but each
+block in Stacks cannot have a parent that was mined more than 100 Bitcoin
+blocks ago.  This is a proposed change to the system.
+
+This latter design point is required to ensure that Stackers can redeem sBTC
+peg-outs in a timely manner while preserving peg safety.  If Stacks can't fork
+beyond 100 Bitcoin blocks, then a peg-out request cannot be invalidated by
+Stacks miners after 100 Bitcoin blocks have passed.  While limiting the fork
+length is necessary to implement sBTC, it introduces some challenges to
+preserving system liveness under adverse conditions: dealing with the threat of
+a persistent 51% miner, and dealing with the threat of a miner that does not
+disclose blocks.
+
+The reader will recall that Stacks miners produce blocks by first sending a
+block-commit transaction to the Bitcoin blockchain, which contains the hash of
+the Stacks block.  However, because Stacks is a single-leader blockchain that
+can fork, there is no protocol-level requirement for the Stacks miner to ever
+broadcast their block to the Stacks network (there are, however, incentives to
+do so -- doing so is the only way in which the miner can receive a block
+reward).  What this means is that a malicious Stacks miner can effectively
+decide the set of chain tips if they win 100 blocks in a row, and halt the
+Stacks chain by winning 100 sortitions in a row and never announcing any Stacks
+blocks.
+
+This proposal addresses both scenarios in three ways:
+
+* We increase the
+economic cost to produce forks after a short time horizon.  To produce a fork
+of at least 7 blocks deep, the miner must have the explicit support of a
```

Comment:
Would it be possible to further reduce the possible fork length? So that even the creation of a 'sibling' would get punished by Stackers? (to further reduce the 51% mining formation risk) 

What is the reason for 7 blocks? Why not 5, 1 or 11?

Reply:
I recall that 7 came from the general people consensus that 6 Bitcoin confirmations equate to finality.

Reply:
We could make the mining field more efficient and fair, if stackers immediately bless the first block in the event of a sibling? What speaks against it?
This would further decrease orphaned blocks produced by bad actors?

Reply:
I don't understand the Bitcoin confirmation argument because a fork on Stacks is different to a fork on Bitcoin. Stack forks happens on network delays or on bad actors. Bitcoin reorgs happens if two blocks are mined at the same time. This is not possible in Stacks.

Reply:
Stackers will be able to vote on thawing blocks with [6,150] confirmations to steer, but not completely control, block production.

We are retiring the "blessing" terminology and replacing it with freezing and thawing.

Reply:
Unfortunately, this doesn´t solve the problem with siblings and forks <6 blocks like we see them right now. E.g. Stacks block 95283/95284?

Reply:
The economic profit on 1 block forks can be high enough to encourage miners to do it.

I understand that stackers should not control block production, however, indicating their preferred fork as soon as possible could be integrated more into the consensus algorithm.

Reply:
Correct. If our goal is to further decentralize the network - wouldn't that be the next necessary step?

Otherwise, larger miners would be able to "bleed out" smaller ones by initiating forks/siblings on purpose?

Reply:
> The economic profit on 1 block forks can be high enough to encourage miners to do it.

Right, the reason why this doesn't happen on Bitcoin (at least in theory) is because miners have a long term commitment with the Bitcoin network i.e. they purchased ASICs whose only way to monetize is mining Bitcoin for years. They are discouraged to damage the trust in the system that allows them to get a profit on their capital expenditure.

But Stacks is mainly operational expenditure, so miners don't have such commitment. They don't have an up-front cost that forces them to stay mining for long periods of time in order to get a profit.

Reply:
Could the text be extended to explain the calculations of the economical costs for a fork? Will get only frozen blocks the block reward?